### PR TITLE
Added autoencoded image output when run in test mode

### PR DESCRIPTION
**Pull request by dribnet**
Running in test mode was not producing any output, so I added
an initial implementation of Trainer.test(). This simply
autoencodes a batch of inputs, and saves the inputs
as well as the autoencoded outputs.

This can be combined with '--split valid' to test
autoencoding on validation data. To make these tests
more repeatable, shuffling was disabled in test
mode and the images are sort ordered.

Sample output attached.
validation inputs
autoencoded

### CODE REVIEW

**Comment by carpedm20** [comment]
Many thanks! But isn't it more reasonable to add `generate` not `autoencode`?

Looks like you also try some training with the code. I received a comment from authors of the paper suggesting that "using two separate optimizer" but it still doesn't helpful yet (but it could be wrong because only 25k step has been passed. I'm curious how's your training goes?

**Comment by dribnet** [comment]
Well, I'm opinionated on the matter but I find `autoencode` much more informative than `generate`. Across epochs/hyperparamaters it is much easier to be comparing a reconstruction with the same inputs. But more generally - I'm ultimately less concerned with the appearance of random decodes than I am with reconstructions and variations of those on a closed set of operations.

Yes - I've been trying out some training with the code. The image above is from 20k steps and training failed just after that. But I realize the project is new and the model is still being developed/debugged. Once the model is working, my interest is in exploring the latent space of a trained model. Happy to share results / useful code changes as I go along.

(edit: aha, re-read the paper last night. yes, `generate` would probably make sense)
